File: cpp/algorithm/union_find/union_find.cpp
#include <iostream>
#include <vector>

using namespace std;

/*
  以下よりunion findを試してみる。acが取れていることを確認している。
  https://atcoder.jp/contests/atc001/tasks/unionfind_a
*/

class UnionFind {
public:
  UnionFind(int n) : _parent(std::vector<int>(n, -1)) {
  }

  int root(int x) {
    if (_parent[x] < 0) {
      return x;
    }
    _parent[x] = root(_parent[x]);
    return _parent[x];
  }

  void unite(int x, int y) {
    x = root(x);
    y = root(y);

    if (x == y) {
      return;
    }

    if (_parent[x] > _parent[y]) {
      swap(x, y);
    }

    _parent[x] += _parent[y];
    _parent[y] = x;
  }

  bool isSame(int x, int y) {
    return root(x) == root(y);
  }

  int size(int x) {
    return -_parent[root(x)];
  }

  int groupNum() {
    int res = 0;
    for (size_t i = 0; i < _parent.size(); ++i) {
      if (_parent.at(i) < 0) {
        ++res;
      }
    }
    return res;
  }

private:
  vector<int> _parent;
};

int main() {
  int n, q;
  cin >> n >> q;
  UnionFind uf(n);
  for (int i = 0; i < q; ++i) {
    int query, x, y;
    cin >> query >> x >> y;
    if (query == 0) {
      uf.unite(x, y);
    } else {
      cout << (uf.isSame(x, y) ? "Yes" : "No") << endl;
    }
  }
  return 0;
}

File: golang/algorithm/union_find.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

/*
  以下よりunion findを試してみる。acが取れていることを確認している。
  https://atcoder.jp/contests/atc001/tasks/unionfind_a
*/

type UnionFind struct {
	parent []int
}

func NewUnionFind(n int) UnionFind {
	nums := make([]int, 0, n)
	for i := 0; i < n; i++ {
		nums = append(nums, -1)
	}
	return UnionFind{parent: nums}
}

func (uf *UnionFind) FindRoot(x int) int {
	if uf.parent[x] < 0 {
		return x
	}
	uf.parent[x] = uf.FindRoot(uf.parent[x])
	return uf.parent[x]
}

func (uf *UnionFind) Unite(x, y int) {
	x = uf.FindRoot(x)
	y = uf.FindRoot(y)
	if x == y {
		return
	}

	if uf.parent[x] > uf.parent[y] {
		x, y = y, x
	}

	uf.parent[x] += uf.parent[y]
	uf.parent[y] = x
}

func (uf *UnionFind) IsSame(x, y int) bool {
	return uf.FindRoot(x) == uf.FindRoot(y)
}

func (uf *UnionFind) Size(x int) int {
	return -uf.parent[uf.FindRoot(x)]
}

func (uf UnionFind) GroupNum() int {
	var res int
	for i := range uf.parent {
		if uf.parent[i] < 0 {
			res++
		}
	}
	return res
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	nqStr := strings.Split(strings.TrimSpace(scanner.Text()), " ")
	n, _ := strconv.Atoi(nqStr[0])
	q, _ := strconv.Atoi(nqStr[1])
	uf := NewUnionFind(n)
	for i := 0; i < q; i++ {
		scanner.Scan()
		queryStr := strings.Split(strings.TrimSpace(scanner.Text()), " ")
		query := queryStr[0]
		v1, _ := strconv.Atoi(queryStr[1])
		v2, _ := strconv.Atoi(queryStr[2])
		if query == "0" {
			uf.Unite(v1, v2)
		} else {
			if uf.IsSame(v1, v2) {
				fmt.Println("Yes")
			} else {
				fmt.Println("No")
			}
		}
	}
}
